Use StyleColor object instead of Color in WebCore

mdubet opened the pull request:
#### b6c018250b04a3642c07f2bd1feb7ad3aaca312a
<pre>
Use StyleColor object instead of Color in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=234691">https://bugs.webkit.org/show_bug.cgi?id=234691</a>
rdar://problem/87050763

Reviewed by Antti Koivisto.

With recent dynamic color features in CSS levels, such as currentcolor or relative color,
WebCore need to propagate an unresolved object which will eventually (at render time)
compute to an actual platform Color.

This patch contains:
  - the implementation of StyleColor, which now represented a
  maybe unresolved yet Color.
  - the specificity of setting the color property (where currentcolor
  is actually the inherited computed value of color)
  is now handled in StyleBuilderCustom.h instead of
  generated by the makeprop.pl script.
  - missing color initialization in object constructors
  (such as BorderValue) are fixed.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::PropertyWrapperColor::PropertyWrapperColor):
(WebCore::PropertyWrapperColor::value const):
(WebCore::PropertyWrapperColor::unresolvedValue const):
(WebCore::PropertyWrapperVisitedAffectedColor::PropertyWrapperVisitedAffectedColor):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
(WebCore::PropertyWrapperMaybeInvalidColor::PropertyWrapperMaybeInvalidColor): Deleted.
(WebCore::PropertyWrapperMaybeInvalidColor::value const): Deleted.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::currentColorOrValidColor const):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::ComputedStyleExtractor::adjustSVGPaint const):
(WebCore::ComputedStyleExtractor::svgPropertyValue):
(WebCore::ComputedStyleExtractor::adjustSVGPaintForCurrentColor const): Deleted.
* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::currentColor):
(WebCore::serializationForRenderTreeAsText):
(WebCore::serializationForCSS):
(WebCore::StyleColor::colorFromAbsoluteKeyword):
(WebCore::StyleColor::colorFromKeyword):
(WebCore::operator&lt;&lt;):
(WebCore::StyleColor::debugDescription const):
(WebCore::StyleColor::resolveColor const):
(WebCore::StyleColor::resolveColorWithoutCurrentColor const):
(WebCore::StyleColor::isCurrentColor const):
(WebCore::StyleColor::isAbsoluteColor const):
(WebCore::StyleColor::absoluteColor const):
(WebCore::StyleColor::StyleColor):
* Source/WebCore/css/StyleColor.h:
* Source/WebCore/css/makeprop.pl:
(generateInheritValueSetter):
(generateValueSetter):
(handleCurrentColorValue): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::removeTextFillAndStrokeColorsIfNeeded):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::CaretBase::computeCaretColor):
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::DataDetection::detectContentInRange):
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::insertDroppedImagePlaceholdersAtCaret):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::calculateHighlightColor const):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustButtonLikeControlStyle const):
* Source/WebCore/rendering/RenderTreeAsText.cpp:
(WebCore::RenderTreeAsText::writeRenderObject):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
* Source/WebCore/rendering/style/BorderValue.cpp:
(WebCore::BorderValue::BorderValue):
(WebCore::BorderValue::isTransparent const):
* Source/WebCore/rendering/style/BorderValue.h:
(WebCore::BorderValue::setColor):
(WebCore::BorderValue::color const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::unresolvedColorForProperty const):
(WebCore::RenderStyle::currentColor):
(WebCore::RenderStyle::isCurrentColor):
(WebCore::RenderStyle::colorResolvingCurrentColor const):
(WebCore::RenderStyle::resolvedColorAndColorFilter const):
(WebCore::RenderStyle::initialTapHighlightColor):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::tapHighlightColor const):
(WebCore::RenderStyle::setBackgroundColor):
(WebCore::RenderStyle::setBorderLeftColor):
(WebCore::RenderStyle::setBorderRightColor):
(WebCore::RenderStyle::setBorderTopColor):
(WebCore::RenderStyle::setBorderBottomColor):
(WebCore::RenderStyle::setOutlineColor):
(WebCore::RenderStyle::setTextStrokeColor):
(WebCore::RenderStyle::setTextFillColor):
(WebCore::RenderStyle::setCaretColor):
(WebCore::RenderStyle::setAccentColor):
(WebCore::RenderStyle::setColumnRuleColor):
(WebCore::RenderStyle::setTextDecorationColor):
(WebCore::RenderStyle::setTextEmphasisColor):
(WebCore::RenderStyle::setTapHighlightColor):
(WebCore::RenderStyle::strokeColor const):
(WebCore::RenderStyle::setStrokeColor):
(WebCore::RenderStyle::setVisitedLinkStrokeColor):
(WebCore::RenderStyle::visitedLinkStrokeColor const):
(WebCore::RenderStyle::initialStrokeColor):
(WebCore::RenderStyle::fillPaintColor const):
(WebCore::RenderStyle::setFillPaintColor):
(WebCore::RenderStyle::strokePaintColor const):
(WebCore::RenderStyle::setStrokePaintColor):
(WebCore::RenderStyle::setStopColor):
(WebCore::RenderStyle::setFloodColor):
(WebCore::RenderStyle::setLightingColor):
(WebCore::RenderStyle::backgroundColorEqualsToColorIgnoringVisited const):
(WebCore::RenderStyle::initialTextStrokeColor):
(WebCore::RenderStyle::initialTextDecorationColor):
(WebCore::RenderStyle::initialBackgroundColor):
(WebCore::RenderStyle::initialTextEmphasisColor):
(WebCore::RenderStyle::setVisitedLinkBackgroundColor):
(WebCore::RenderStyle::setVisitedLinkBorderLeftColor):
(WebCore::RenderStyle::setVisitedLinkBorderRightColor):
(WebCore::RenderStyle::setVisitedLinkBorderBottomColor):
(WebCore::RenderStyle::setVisitedLinkBorderTopColor):
(WebCore::RenderStyle::setVisitedLinkOutlineColor):
(WebCore::RenderStyle::setVisitedLinkColumnRuleColor):
(WebCore::RenderStyle::setVisitedLinkTextDecorationColor):
(WebCore::RenderStyle::setVisitedLinkTextEmphasisColor):
(WebCore::RenderStyle::setVisitedLinkTextFillColor):
(WebCore::RenderStyle::setVisitedLinkTextStrokeColor):
(WebCore::RenderStyle::setVisitedLinkCaretColor):
(WebCore::RenderStyle::borderLeftColor const):
(WebCore::RenderStyle::borderRightColor const):
(WebCore::RenderStyle::borderTopColor const):
(WebCore::RenderStyle::borderBottomColor const):
(WebCore::RenderStyle::backgroundColor const):
(WebCore::RenderStyle::columnRuleColor const):
(WebCore::RenderStyle::outlineColor const):
(WebCore::RenderStyle::textEmphasisColor const):
(WebCore::RenderStyle::textFillColor const):
(WebCore::RenderStyle::initialTextFillColor):
(WebCore::RenderStyle::textStrokeColor const):
(WebCore::RenderStyle::caretColor const):
(WebCore::RenderStyle::visitedLinkBackgroundColor const):
(WebCore::RenderStyle::visitedLinkBorderLeftColor const):
(WebCore::RenderStyle::visitedLinkBorderRightColor const):
(WebCore::RenderStyle::visitedLinkBorderBottomColor const):
(WebCore::RenderStyle::visitedLinkBorderTopColor const):
(WebCore::RenderStyle::visitedLinkOutlineColor const):
(WebCore::RenderStyle::visitedLinkColumnRuleColor const):
(WebCore::RenderStyle::textDecorationColor const):
(WebCore::RenderStyle::visitedLinkTextDecorationColor const):
(WebCore::RenderStyle::visitedLinkTextEmphasisColor const):
(WebCore::RenderStyle::visitedLinkTextFillColor const):
(WebCore::RenderStyle::visitedLinkTextStrokeColor const):
(WebCore::RenderStyle::visitedLinkCaretColor const):
(WebCore::RenderStyle::stopColor const):
(WebCore::RenderStyle::floodColor const):
(WebCore::RenderStyle::lightingColor const):
(WebCore::RenderStyle::accentColor const):
(WebCore::RenderStyle::currentColor): Deleted.
(WebCore::RenderStyle::isCurrentColor): Deleted.
* Source/WebCore/rendering/style/SVGRenderStyle.h:
(WebCore::SVGRenderStyle::initialFillPaintColor):
(WebCore::SVGRenderStyle::initialStrokePaintColor):
(WebCore::SVGRenderStyle::initialStopColor):
(WebCore::SVGRenderStyle::initialFloodColor):
(WebCore::SVGRenderStyle::initialLightingColor):
(WebCore::SVGRenderStyle::fillPaintColor const):
(WebCore::SVGRenderStyle::strokePaintColor const):
(WebCore::SVGRenderStyle::stopColor const):
(WebCore::SVGRenderStyle::floodColor const):
(WebCore::SVGRenderStyle::lightingColor const):
(WebCore::SVGRenderStyle::visitedLinkFillPaintColor const):
(WebCore::SVGRenderStyle::visitedLinkStrokePaintColor const):
(WebCore::SVGRenderStyle::setFillPaint):
(WebCore::SVGRenderStyle::setStrokePaint):
(WebCore::SVGRenderStyle::setStopColor):
(WebCore::SVGRenderStyle::setFloodColor):
(WebCore::SVGRenderStyle::setLightingColor):
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/rendering/style/StyleMultiColData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/RenderSVGResource.cpp:
(WebCore::inheritColorFromParentStyleIfNeeded):
(WebCore::requestPaintingResource):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTapHighlightColor):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritCaretColor):
(WebCore::Style::colorAndSVGPaintType):
(WebCore::Style::BuilderCustom::applyValueColor):
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::colorFromPrimitiveValue const):
(WebCore::Style::BuilderState::colorFromPrimitiveValueWithResolvedCurrentColor const):
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
(WebCore::SVGFEDiffuseLightingElement::setFilterEffectAttribute):
(WebCore::SVGFEDiffuseLightingElement::createFilterEffect const):
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
(WebCore::SVGFEDropShadowElement::setFilterEffectAttribute):
(WebCore::SVGFEDropShadowElement::createFilterEffect const):
* Source/WebCore/svg/SVGFEFloodElement.cpp:
(WebCore::SVGFEFloodElement::setFilterEffectAttribute):
(WebCore::SVGFEFloodElement::createFilterEffect const):
* Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
(WebCore::SVGFESpecularLightingElement::setFilterEffectAttribute):
(WebCore::SVGFESpecularLightingElement::createFilterEffect const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::sendTapHighlightForNodeIfNecessary):

Canonical link: <a href="https://commits.webkit.org/255095@main">https://commits.webkit.org/255095@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad6098d854a5a014828f7044cc53c457327182f9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/310 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101056 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/160984 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95315 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/322 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83673 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97436 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96967 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/27249 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82203 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70276 "Updated gtk dependencies (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35448 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/15902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33245 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16987 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3541 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37033 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38956 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36096 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -319,7 +319,7 @@ URL ApplicationManifestParser::parseScope(const JSON::Object& manifest, const UR
 Color ApplicationManifestParser::parseColor(const JSON::Object& manifest, const String& propertyName)
 {
-    return CSSParser::parseColorWithoutContext(parseGenericString(manifest, propertyName));
+    return (Color)CSSParser::parseColorWithoutContext(parseGenericString(manifest, propertyName));
```
C-style casts like these are not kosher in the WebKit coding style. If it needs to be a cast, it should be a static_cast, but I think replacing the `operator Color` on StyleColor with a function would be more idiomatic and clearer.

mdubet replied:
Yes I will remove the conversion operator and use a regular well named function.

weinig commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +30,119 @@
+    std::optional<Color> currentColor = std::nullopt;
+};
+
+struct ColorMixPercentages {
```
I would take the opportunity to put this and the other ColorMix related structs / functions in their own ColorMix.h/cpp

weinig commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +30,119 @@
+    Color computeColor(ColorContext = { }) const;
+    bool isValid() const;
+
+    WEBCORE_EXPORT explicit operator Color() const;
```
I would prefer to avoid adding the `operator Color()` if possible, as it often makes refactoring harder due to it being marginally harder to find calls.

weinig commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +30,119 @@
     static bool isColorKeyword(CSSValueID, OptionSet<CSSColorType> = { CSSColorType::Absolute, CSSColorType::Current, CSSColorType::System });
+
+    StyleColor() = default;
+    StyleColor(Color const&);
```
You have this in quite a few places, but WebKit style is to write the `const` first, so `const Color&`

nt1m replied:
@mdubet This doesn't seem fixed

weinig commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -383,14 +383,14 @@ String StyleProperties::getPropertyValue(CSSPropertyID propertyID, Document* doc
 
     auto& primitiveColor = downcast<CSSPrimitiveValue>(*colorValue);
-    return primitiveColor.isRGBColor() ? primitiveColor.color() : CSSParser::parseColorWithoutContext(colorValue->cssText());
+    return primitiveColor.isRGBColor() ? primitiveColor.color() : StyleColor { CSSParser::parseColorWithoutContext(colorValue->cssText()) };
```
I think parseColorWithoutContext() returns a StyleColor, so the wrapping is unneeded.

weinig commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -1177,8 +1177,7 @@ struct LengthOrPercentRawConsumer {
         PercentRawKnownTokenTypeFunctionConsumer,
-        LengthRawKnownTokenTypeFunctionConsumer
-    >;
+        LengthRawKnownTokenTypeFunctionConsumer>;
```
This doesn't seem like it is part of the change. Should leave off these unrelated style or whitespace changes.

weinig commented on Source/WebCore/rendering/style/BorderValue.h:
```diff
@@ -24,29 +24,23 @@
-    {
-        return m_color.isValid() && !m_color.isVisible();
-    }
+    bool isTransparent() const;
```
We probably need to revisit this. This likely should be processed like all other colors (e.g. visitedDependentColor...). Not a change from status quo, so no need to do anything, but maybe worth a FIXME and bugzilla.

pxlcoder commented on Source/WebCore/css/CSSValuePool.cpp:
```diff
@@ -40,9 +40,9 @@ StaticCSSValuePool::StaticCSSValuePool()
-    m_blackColor.construct(CSSValue::StaticCSSValue, Color::black);
+    m_transparentColor.construct(CSSValue::StaticCSSValue, StyleColor{Color::transparentBlack});
+    m_whiteColor.construct(CSSValue::StaticCSSValue, StyleColor{Color::white});
+    m_blackColor.construct(CSSValue::StaticCSSValue, StyleColor{Color::black});
```
Nit: Formatting: `StyleColor { Color::transparentBlack }`

pxlcoder commented on Source/WebCore/css/StyleColor.cpp:
```diff
@@ -32,19 +32,58 @@
+        return "currentcolor"_s;
+
+    return "color-mix()"_s;
+    */
```
Commented code should be removed.

pxlcoder commented on Source/WebCore/css/StyleColor.cpp:
```diff
@@ -32,19 +32,58 @@
+
+Color StyleColor::colorFromKeyword(CSSValueID keyword , OptionSet<StyleColorOptions> options)
+{
+    if(isAbsoluteColorKeyword(keyword))
```
Nit: Spacing: `if (isAbsoluteColorKeyword(keyword))`

pxlcoder commented on Source/WebCore/css/StyleColor.cpp:
```diff
@@ -81,4 +136,169 @@ bool StyleColor::isColorKeyword(CSSValueID id, OptionSet<CSSColorType> allowedCo
+Color StyleColor::computeColor(ColorContext context) const
+{
+    if (isAbsoluteColor())
+        return this->absoluteColor();
```
I don't think `this->` is necessary here. There are a couple of other instances in this class too.

pxlcoder commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +30,119 @@
+    friend WTF::TextStream& operator<<(WTF::TextStream&, ColorMixPercentages const&);
+};
+
+struct ColorMix {
```
I think this should probably be a `class`, since it has private members.

pxlcoder commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +30,119 @@
+    String debugDescription() const;
+
+private:
+    ColorKind m_color = Color { };
```
Explicit initialization is not necessary, since `std::variant` will use the first alternative.

pxlcoder commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +30,119 @@
+private:
+    ColorKind m_color = Color { };
+    // the color property needs to remember its inherit color at runtime.
+    std::optional<Color> m_inheritedColor = { };
```
Explicit initialization is not necessary, this will be default initialized to `std::nullopt`.

pxlcoder commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +30,119 @@
+
+private:
+    ColorKind m_color = Color { };
+    // the color property needs to remember its inherit color at runtime.
```
This comment doesn't really add useful information, since the statement is implied by the fact that this member exists. The "Why?" would be more useful.

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -872,13 +872,13 @@ static Color normalizedColor(Color color, bool ignoreDefaultColor, Element& elem
     if (RefPtr<CSSValue> value = computedStylePropertyForElement(element, propertyId)) {
         if (is<CSSPrimitiveValue>(*value) && downcast<CSSPrimitiveValue>(*value).isRGBColor())
-            return normalizedColor(downcast<CSSPrimitiveValue>(*value).color(), ignoreDefaultColor, element);
+            return normalizedColor((Color)downcast<CSSPrimitiveValue>(*value).color(), ignoreDefaultColor, element);
```
This could be rewritten to use `dynamicDowncast` and variable assignment in the conditional.

`if(auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value); primitiveValue && primitiveValue->isRGBColor())`

You can then use `primitiveValue` in the return statement.

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -872,13 +872,13 @@ static Color normalizedColor(Color color, bool ignoreDefaultColor, Element& elem
 
     bool inherit = false;
     if (RefPtr<CSSValue> value = inlineStylePropertyForElement(element, propertyId)) {
         if (is<CSSPrimitiveValue>(*value) && downcast<CSSPrimitiveValue>(*value).isRGBColor())
```
Same comment about `dynamicDowncast`.

pxlcoder commented on Source/WebCore/html/canvas/CanvasStyle.cpp:
```diff
@@ -46,6 +47,7 @@
 
 namespace WebCore {
 
+// FIXME why is everything related to Color reimplement here?
```
If we intend to keep this `FIXME`, let's rewrite as `FIXME:  Why is everything related to Color reimplemented here?`

pxlcoder commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -1495,7 +1495,8 @@ void DragController::insertDroppedImagePlaceholdersAtCaret(const Vector<IntSize>
         image->setAttributeWithoutSynchronization(HTMLNames::heightAttr, AtomString::number(size.height()));
         image->setInlineStyleProperty(CSSPropertyMaxWidth, 100, CSSUnitType::CSS_PERCENTAGE);
-        image->setInlineStyleProperty(CSSPropertyBackgroundColor, serializationForCSS(Color::black.colorWithAlphaByte(13)));
+        Color backgroundColor = Color::black.colorWithAlphaByte(13);
```
Is this temporary variable necessary?

pxlcoder commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -929,11 +932,11 @@ inline void BuilderCustom::applyTextOrBoxShadowValue(BuilderState& builderState,
         auto spread = shadowValue.spread ? shadowValue.spread->computeLength<Length>(conversionData) : Length(0, LengthType::Fixed);
         ShadowStyle shadowStyle = shadowValue.style && shadowValue.style->valueID() == CSSValueInset ? ShadowStyle::Inset : ShadowStyle::Normal;
-        Color color;
+        Color color; // FIXME should be resolved later.
```
Same comment about `FIXME` style as above.

anttijk commented on Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
```diff
@@ -465,12 +465,11 @@ static RefPtr<CSSValue> positionOffsetValue(const RenderStyle& style, CSSPropert
-    if (!color.isValid())
-        return CSSValuePool::singleton().createColorValue(style->color());
-    return CSSValuePool::singleton().createColorValue(color);
+    auto resolvedColor = style->colorResolvingCurrentColor(color, false);
```
'false' here is mysterious.

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +31,86 @@
+private:
+    ColorKind m_color;
+    // The color property may depends on its inherited value at runtime.
+    std::optional<Color> m_inheritedColor;
```
There shouldn't be any need for this field. The colors are resolved at use time. You can see there is nothing like this field at the moment and we manage ok. 

As said before, the first patch should just introduce StyleColor as a direct replacement for Color in RenderStyle, encapsulating a color value and 'current color' state. No new functionality should be included, this is complex enough change in itself. Internally it can even keep using null Color to signify 'current color' so there is no memory or performance concerns. In future patches we can look into what is the best way to implement new functionality on top.

mdubet replied:
This is a leftover from the initial patch (with the color-mix feature) indeed.

anttijk commented on Source/WebCore/rendering/style/BorderValue.cpp:
```diff
@@ -0,0 +1,50 @@
+
+namespace WebCore {
+
+BorderValue::BorderValue()
```
You could land this sort of easily reviewed minor refactorings (moving code to cpp here) separately.

mdubet replied:
It landed in another PR ✔️

anttijk commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -2988,14 +2967,37 @@ static Color parseColorFunction(CSSParserTokenRange& range, const CSSParserConte
+    // Return a valid style color.
+    range = colorRange;
     return color;
 }
```
Anything to do with color-mix should be done separately.

mdubet replied:
Yes this was a leftover from the previous unified patch

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -31,31 +31,86 @@
+    void setInheritedColor(const Color&);
+    std::optional<Color> inheritedColor() const;
+
+    WEBCORE_EXPORT Color resolveColor(ColorContext = { }) const;
```
How would callers know when to pass a ColorContext and when not (since it is optional)?

mdubet replied:
I've added a comment to explain the purpose of this. Normally you should always provide a context, but the state of the codebase makes it impossible sometimes.
Maybe we should remove the optional value and make it explicit that you pass an empty context ?

anttijk commented on LayoutTests/TestExpectations:
```diff
@@ -6003,3 +6003,7 @@ js/ShadowRealm-iframe-detach.html [ Pass Failure ]
+
+# Old SVG tests with invalid results in SVG2 + Color 4
+imported/w3c/web-platform-tests/svg/import/color-prop-05-t-manual.svg [ Pass Failure ]
+svg/W3C-SVG-1.1-SE/color-prop-05-t.svg [ Pass Failure ]
```
This patch should not contain any functional changes and so shouldn't affect any tests.

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -521,27 +522,26 @@ void EditingStyle::init(Node* node, PropertiesToInclude propertiesToInclude)
-                m_mutableStyle->setProperty(CSSPropertyFontSize, cssValue->cssText());
+    if (node) {
+        auto renderStyle = node->computedStyle();
+        if (renderStyle) {
```
```suggestion
        if (auto* renderStyle = node->computedStyle()) {
```

nt1m commented on Source/WebCore/rendering/style/BorderValue.cpp:
```diff
@@ -25,23 +25,26 @@
-    , m_isAuto(static_cast<unsigned>(OutlineIsAuto::Off))
+: m_color { CurrentColor::make() }
+, m_style { static_cast<unsigned>(BorderStyle::None) }
+, m_isAuto { static_cast<unsigned>(OutlineIsAuto::Off) }
```
nit: please preserve indentation

anttijk commented on Source/WebCore/css/CSSPrimitiveValue.h:
```diff
@@ -177,7 +178,7 @@ class CSSPrimitiveValue final : public CSSValue {
     WEBCORE_EXPORT String stringValue() const;
 
-    const Color& color() const { ASSERT(primitiveUnitType() == CSSUnitType::CSS_RGBCOLOR); return *m_value.color; }
+    const StyleColor& color() const { ASSERT(primitiveUnitType() == CSSUnitType::CSS_RGBCOLOR); return *m_value.color; }
```
StyleColor represents a post-style resolution (computed style) color in RenderStyle. It shouldn't be used in CSSValue (which is pre-style resolution). 'Currentcolor' on this level is represented by id value CSSValueCurrentcolor so a concrete color value works just fine at least for now. The mapping to StyleColor happens in Style::Builder and its helpers.

I don't think this patch needs to touch CSSPrimitiveValue at all. Making this change should make this patch way more tractable.

mdubet replied:
I'm influenced by the fact that I know we will need this very soon, but for the purpose of this patch indeed it's not necessary. I'm gonna remove this for the patch.

anttijk commented on Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
```diff
@@ -86,7 +86,8 @@ bool SVGFEDiffuseLightingElement::setFilterEffectAttribute(FilterEffect& effect,
         ASSERT(renderer);
-        Color color = renderer->style().colorByApplyingColorFilter(renderer->style().svgStyle().lightingColor());
+        auto& style = renderer->style();
+        Color color = style.colorByApplyingColorFilter(style.colorResolvingCurrentColor(style.svgStyle().lightingColor()));
```
You could have a helper for this (resolveColorAndApplyColorFilter or something) that takes StyleColor.

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -34,21 +34,42 @@
+    std::optional<Color> currentColor = std::nullopt;
+};
+
+using ColorKind = std::variant<Color, CurrentColor, SystemColor>;
```
These are internal details of StyleColor and would better be scoped to it. 

I don't see any use for SystemColor (which carries no data anyway)

ColorContext type doesn't seem to add anything. It could be replaced by simply passing Color.

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -60,8 +81,38 @@ struct StyleColor {
+    // if you don't pass it, the resolved color might be innacurate.
+    // In some part of the codebase, we don't have enough information
+    // to pass the correct context yet.
+    WEBCORE_EXPORT Color resolveColor(ColorContext = { }) const;
```
Nothing is being resolved without a context. I think clients that don't provide context should just use absoluteColor(). If we need to deal with cases where we have value 'currentcolor' but are not currently resolving it correctly then there should be a separate helper that communicated this (like absoluteOrInvalidColorIgnoringCurrentColor()) with a FIXME to fix the clients and remove the function.

The context argument (which can be just plain Color) should be mandatory.

anttijk commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -536,12 +536,9 @@ void EditingStyle::init(Node* node, PropertiesToInclude propertiesToInclude)
+    if (RenderStyle::isCurrentColor(renderStyle->textFillColor()))
         m_mutableStyle->removeProperty(CSSPropertyWebkitTextFillColor);
-    if (!renderStyle->textStrokeColor().isValid())
+    if (RenderStyle::isCurrentColor(renderStyle->textStrokeColor()))
```
This can just test renderStyle->textStrokeColor().isCurrentColor().
RenderStyle::isCurrentColor can be removed (now or later).

mdubet replied:
Yes `RenderStyle::isCurrentColor` will be removed in a future patch.

anttijk commented on Source/WebCore/rendering/style/BorderValue.cpp:
```diff
@@ -25,23 +25,26 @@
 {
-    return nonZero() && !isTransparent() && style() != BorderStyle::Hidden;
+    return nonZero() && !isTransparent(renderStyle) && style() != BorderStyle::Hidden;
 }
```
Let's not do this (and the large associated refactoring) in this patch. Passing RenderStyle like this is not nice and you again run into question of when is it ok not to pass it.  

You can just use absoluteOrInvalidColorIgnoringCurrentColor() mentioned elsewhere.

This presumable fixes a bug so there should be associated test too.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2279,20 +2283,26 @@ Color RenderStyle::colorResolvingCurrentColor(CSSPropertyID colorProperty, bool
-    if (isCurrentColor(color))
-        return this->color();
+    ColorContext colorContext;
+    colorContext.currentColor = this->color().resolveColor();
```
This should be able to use absoluteColor() as 'currentcolor' for color property gets translated in style builder.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2279,20 +2283,26 @@ Color RenderStyle::colorResolvingCurrentColor(CSSPropertyID colorProperty, bool
-    return color;
+Color RenderStyle::colorCurrentColor(bool visitedLink) const
+{
+    return colorResolvingCurrentColor(visitedLink ? visitedLinkColor() : color());
```
color() should never be currentcolor in RenderStyle so this function should not be needed (and should not be doing anything).

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -34,21 +34,42 @@
 
 namespace WebCore {
 
+class ColorMixComponent;
```
Some ColorMix

anttijk commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -2040,6 +2044,30 @@ inline void BuilderCustom::applyValueStrokeColor(BuilderState& builderState, CSS
+            return parentStyle.color();
+        }
+        return color;
+    };
```
Here we resolve currentcolor for color property, making the other bits in the patch where it is resolved again unnecessary.

This sort of point out that the color and visitedLinkColor should remain as Color instead of StyleColor. That would clean up other parts of the patch.

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -60,8 +81,38 @@ struct StyleColor {
+
+    StyleColor();
+    StyleColor(const Color&);
+    StyleColor(const ColorKind&);
```
ColorKind should be an internal detail, and this constructor private . You could have for example

static StyleColor StyleColor::currentColor()  

for constructing the values instead.

nt1m commented on Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
```diff
@@ -160,8 +161,9 @@ RefPtr<FilterEffect> SVGFEDiffuseLightingElement::createFilterEffect(const Filte
+    auto& style = renderer->style();
 
-    Color color = renderer->style().colorByApplyingColorFilter(renderer->style().svgStyle().lightingColor());
+    Color color = style.colorByApplyingColorFilter(style.colorResolvingCurrentColor(style.svgStyle().lightingColor()));
```
you can use your new `resolvedColorByApplyingColorFilter`

nt1m commented on Source/WebCore/svg/SVGFEDropShadowElement.cpp:
```diff
@@ -157,9 +157,10 @@ RefPtr<FilterEffect> SVGFEDropShadowElement::createFilterEffect(const FilterEffe
+    const SVGRenderStyle& svgStyle = style.svgStyle();
     
-    Color color = renderer->style().colorByApplyingColorFilter(svgStyle.floodColor());
+    Color color = style.colorByApplyingColorFilter(style.colorResolvingCurrentColor(svgStyle.floodColor()));
```
ditto

nt1m commented on Source/WebCore/svg/SVGFEFloodElement.cpp:
```diff
@@ -65,9 +65,10 @@ RefPtr<FilterEffect> SVGFEFloodElement::createFilterEffect(const FilterEffectVec
+    const SVGRenderStyle& svgStyle = style.svgStyle();
 
-    Color color = renderer->style().colorByApplyingColorFilter(svgStyle.floodColor());
+    auto color = style.colorByApplyingColorFilter(style.colorResolvingCurrentColor(svgStyle.floodColor()));
```
same here

nt1m commented on Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
```diff
@@ -94,7 +94,8 @@ bool SVGFESpecularLightingElement::setFilterEffectAttribute(FilterEffect& effect
         ASSERT(renderer);
-        Color color = renderer->style().colorByApplyingColorFilter(renderer->style().svgStyle().lightingColor());
+        auto& style = renderer->style();
+        auto color = style.colorByApplyingColorFilter(style.colorResolvingCurrentColor(style.svgStyle().lightingColor()));
```
same here

nt1m commented on Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
```diff
@@ -169,8 +170,9 @@ RefPtr<FilterEffect> SVGFESpecularLightingElement::createFilterEffect(const Filt
+    auto& style = renderer->style();
 
-    Color color = renderer->style().colorByApplyingColorFilter(renderer->style().svgStyle().lightingColor());
+    auto color = style.colorByApplyingColorFilter(style.colorResolvingCurrentColor(style.svgStyle().lightingColor()));
```
same here

nt1m commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -60,8 +72,38 @@ struct StyleColor {
+
+    // The context is necessary to be able to resolve "current color",
+    // if you don't pass it, the resolved color might be innacurate.
+    // In some part of the codebase, we don't have enough information
```
nit: some parts

nt1m commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2335,6 +2337,11 @@ Color RenderStyle::colorByApplyingColorFilter(const Color& color) const
     return transformedColor;
 }
 
+Color RenderStyle::resolvedColorByApplyingColorFilter(const StyleColor& color) const
```
this name is a bit confusing, it suggests that we're resolving the color by using the color filter, rather than, resolving the color then applying the color filter.

Maybe `resolvedColorWithColorFilter` would be more accurate or what Antti originally suggested. Also we're using `resolve` vs. `resolved` in places for colors, would be nice to be consistent.

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -60,8 +72,38 @@ struct StyleColor {
+    // if you don't pass it, the resolved color might be innacurate.
+    // In some part of the codebase, we don't have enough information
+    // to pass the correct context yet.
+    WEBCORE_EXPORT Color resolveColor(std::optional<Color> colorPropertyValue) const;
```
I don't think the argument should be optional. 

If needed, there should be a separate function that returns invalid Color for 'currentcolor' (matching the current behavior of such clients) and explains with its name what it does, colorWithoutResolvingCurrentColor() or something.

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -60,8 +72,38 @@ struct StyleColor {
+
+    static StyleColor currentColor();
+    StyleColor();
+    StyleColor(const Color&);
```
Might want to mark this explicit. I think it reveals some cases where Color is unnecessarily converted to StyleColor. (or maybe it just creates noise, hard to tell).

anttijk commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1229,11 +1229,11 @@ static void adjustInputElementButtonStyle(RenderStyle& style, const HTMLInputEle
         return;
 
-    auto textColor = style.color();
+    auto textColor = style.colorResolvingCurrentColor(style.color());
```
Calling colorResolvingCurrentColor does nothing here since style.color() is already Color. This would be revealed by making StyleColor(Color) constructor explicit.

anttijk commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1229,11 +1229,11 @@ static void adjustInputElementButtonStyle(RenderStyle& style, const HTMLInputEle
 
-    auto textColor = style.color();
+    auto textColor = style.colorResolvingCurrentColor(style.color());
     if (textColor.isValid())
```
Not from this patch, but it shouldn't ever be invalid either.

anttijk commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -60,8 +72,38 @@ struct StyleColor {
+    static StyleColor currentColor();
+    StyleColor();
+    StyleColor(const Color&);
+    StyleColor(const StyleColor&) = default;
```
could add move constructor too StyleColor(StyleColor&&)

nt1m commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1848,11 +1848,12 @@ Color CaretBase::computeCaretColor(const RenderStyle& elementStyle, const Node*
     auto* parentStyle = parentElement && parentElement->renderer() ? &parentElement->renderer()->style() : nullptr;
-    if (elementStyle.hasAutoCaretColor() && parentStyle) {
+    // CSS value "auto" is treated as an invalid color.
+    if (!elementStyle.colorResolvingCurrentColor(elementStyle.caretColor()).isValid() && parentStyle) {
```
This seems like a more verbose way to say the same thing, if the check for `isValid()` is needed now, the comment probably needs to be changed. Otherwise, we should do this change once we actually need it.

mdubet replied:
Yes this change is not needed since the rebasing.

nt1m commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2335,6 +2337,11 @@ Color RenderStyle::colorByApplyingColorFilter(const Color& color) const
     return transformedColor;
 }
 
+Color RenderStyle::resolvedColorAndColorFilter(const StyleColor& color) const
```
probably nitpicky about naming here, but `resolvedColorAndColorFilter` sounds like we're resolving the -apple-color-filter property.

`resolveColorAndApplyColorFilter` (what Antti suggested) or `resolveColorWithColorFilter` make more sense here.

mdubet replied:
`resolveColorWithColorFilter` is more confusing than `resolvedColorAndColorFilter` to me.
I will let the name as is for this PR, feel free to change it in another one if it's worth it.

mdubet committed:
Use StyleColor object instead of Color in WebCore
https://bugs.webkit.org/show_bug.cgi?id=234691
rdar://problem/87050763

Reviewed by Antti Koivisto.

With recent dynamic color features in CSS levels, such as currentcolor or relative color,
WebCore need to propagate an unresolved object which will eventually (at render time)
compute to an actual platform Color.

This patch contains:
  - the implementation of StyleColor, which now represented a
  maybe unresolved yet Color.
  - the specificity of setting the color property (where currentcolor
  is actually the inherited computed value of color)
  is now handled in StyleBuilderCustom.h instead of
  generated by the makeprop.pl script.
  - missing color initialization in object constructors
  (such as BorderValue) are fixed.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::PropertyWrapperColor::PropertyWrapperColor):
(WebCore::PropertyWrapperColor::value const):
(WebCore::PropertyWrapperColor::unresolvedValue const):
(WebCore::PropertyWrapperVisitedAffectedColor::PropertyWrapperVisitedAffectedColor):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
(WebCore::PropertyWrapperMaybeInvalidColor::PropertyWrapperMaybeInvalidColor): Deleted.
(WebCore::PropertyWrapperMaybeInvalidColor::value const): Deleted.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::currentColorOrValidColor const):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::ComputedStyleExtractor::adjustSVGPaint const):
(WebCore::ComputedStyleExtractor::svgPropertyValue):
(WebCore::ComputedStyleExtractor::adjustSVGPaintForCurrentColor const): Deleted.
* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::currentColor):
(WebCore::serializationForRenderTreeAsText):
(WebCore::serializationForCSS):
(WebCore::StyleColor::colorFromAbsoluteKeyword):
(WebCore::StyleColor::colorFromKeyword):
(WebCore::operator<<):
(WebCore::StyleColor::debugDescription const):
(WebCore::StyleColor::resolveColor const):
(WebCore::StyleColor::resolveColorWithoutCurrentColor const):
(WebCore::StyleColor::isCurrentColor const):
(WebCore::StyleColor::isAbsoluteColor const):
(WebCore::StyleColor::absoluteColor const):
(WebCore::StyleColor::StyleColor):
* Source/WebCore/css/StyleColor.h:
* Source/WebCore/css/makeprop.pl:
(generateInheritValueSetter):
(generateValueSetter):
(handleCurrentColorValue): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::removeTextFillAndStrokeColorsIfNeeded):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::CaretBase::computeCaretColor):
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::DataDetection::detectContentInRange):
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::insertDroppedImagePlaceholdersAtCaret):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::calculateHighlightColor const):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustButtonLikeControlStyle const):
* Source/WebCore/rendering/RenderTreeAsText.cpp:
(WebCore::RenderTreeAsText::writeRenderObject):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
* Source/WebCore/rendering/style/BorderValue.cpp:
(WebCore::BorderValue::BorderValue):
(WebCore::BorderValue::isTransparent const):
* Source/WebCore/rendering/style/BorderValue.h:
(WebCore::BorderValue::setColor):
(WebCore::BorderValue::color const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::unresolvedColorForProperty const):
(WebCore::RenderStyle::currentColor):
(WebCore::RenderStyle::isCurrentColor):
(WebCore::RenderStyle::colorResolvingCurrentColor const):
(WebCore::RenderStyle::resolvedColorAndColorFilter const):
(WebCore::RenderStyle::initialTapHighlightColor):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::tapHighlightColor const):
(WebCore::RenderStyle::setBackgroundColor):
(WebCore::RenderStyle::setBorderLeftColor):
(WebCore::RenderStyle::setBorderRightColor):
(WebCore::RenderStyle::setBorderTopColor):
(WebCore::RenderStyle::setBorderBottomColor):
(WebCore::RenderStyle::setOutlineColor):
(WebCore::RenderStyle::setTextStrokeColor):
(WebCore::RenderStyle::setTextFillColor):
(WebCore::RenderStyle::setCaretColor):
(WebCore::RenderStyle::setAccentColor):
(WebCore::RenderStyle::setColumnRuleColor):
(WebCore::RenderStyle::setTextDecorationColor):
(WebCore::RenderStyle::setTextEmphasisColor):
(WebCore::RenderStyle::setTapHighlightColor):
(WebCore::RenderStyle::strokeColor const):
(WebCore::RenderStyle::setStrokeColor):
(WebCore::RenderStyle::setVisitedLinkStrokeColor):
(WebCore::RenderStyle::visitedLinkStrokeColor const):
(WebCore::RenderStyle::initialStrokeColor):
(WebCore::RenderStyle::fillPaintColor const):
(WebCore::RenderStyle::setFillPaintColor):
(WebCore::RenderStyle::strokePaintColor const):
(WebCore::RenderStyle::setStrokePaintColor):
(WebCore::RenderStyle::setStopColor):
(WebCore::RenderStyle::setFloodColor):
(WebCore::RenderStyle::setLightingColor):
(WebCore::RenderStyle::backgroundColorEqualsToColorIgnoringVisited const):
(WebCore::RenderStyle::initialTextStrokeColor):
(WebCore::RenderStyle::initialTextDecorationColor):
(WebCore::RenderStyle::initialBackgroundColor):
(WebCore::RenderStyle::initialTextEmphasisColor):
(WebCore::RenderStyle::setVisitedLinkBackgroundColor):
(WebCore::RenderStyle::setVisitedLinkBorderLeftColor):
(WebCore::RenderStyle::setVisitedLinkBorderRightColor):
(WebCore::RenderStyle::setVisitedLinkBorderBottomColor):
(WebCore::RenderStyle::setVisitedLinkBorderTopColor):
(WebCore::RenderStyle::setVisitedLinkOutlineColor):
(WebCore::RenderStyle::setVisitedLinkColumnRuleColor):
(WebCore::RenderStyle::setVisitedLinkTextDecorationColor):
(WebCore::RenderStyle::setVisitedLinkTextEmphasisColor):
(WebCore::RenderStyle::setVisitedLinkTextFillColor):
(WebCore::RenderStyle::setVisitedLinkTextStrokeColor):
(WebCore::RenderStyle::setVisitedLinkCaretColor):
(WebCore::RenderStyle::borderLeftColor const):
(WebCore::RenderStyle::borderRightColor const):
(WebCore::RenderStyle::borderTopColor const):
(WebCore::RenderStyle::borderBottomColor const):
(WebCore::RenderStyle::backgroundColor const):
(WebCore::RenderStyle::columnRuleColor const):
(WebCore::RenderStyle::outlineColor const):
(WebCore::RenderStyle::textEmphasisColor const):
(WebCore::RenderStyle::textFillColor const):
(WebCore::RenderStyle::initialTextFillColor):
(WebCore::RenderStyle::textStrokeColor const):
(WebCore::RenderStyle::caretColor const):
(WebCore::RenderStyle::visitedLinkBackgroundColor const):
(WebCore::RenderStyle::visitedLinkBorderLeftColor const):
(WebCore::RenderStyle::visitedLinkBorderRightColor const):
(WebCore::RenderStyle::visitedLinkBorderBottomColor const):
(WebCore::RenderStyle::visitedLinkBorderTopColor const):
(WebCore::RenderStyle::visitedLinkOutlineColor const):
(WebCore::RenderStyle::visitedLinkColumnRuleColor const):
(WebCore::RenderStyle::textDecorationColor const):
(WebCore::RenderStyle::visitedLinkTextDecorationColor const):
(WebCore::RenderStyle::visitedLinkTextEmphasisColor const):
(WebCore::RenderStyle::visitedLinkTextFillColor const):
(WebCore::RenderStyle::visitedLinkTextStrokeColor const):
(WebCore::RenderStyle::visitedLinkCaretColor const):
(WebCore::RenderStyle::stopColor const):
(WebCore::RenderStyle::floodColor const):
(WebCore::RenderStyle::lightingColor const):
(WebCore::RenderStyle::accentColor const):
(WebCore::RenderStyle::currentColor): Deleted.
(WebCore::RenderStyle::isCurrentColor): Deleted.
* Source/WebCore/rendering/style/SVGRenderStyle.h:
(WebCore::SVGRenderStyle::initialFillPaintColor):
(WebCore::SVGRenderStyle::initialStrokePaintColor):
(WebCore::SVGRenderStyle::initialStopColor):
(WebCore::SVGRenderStyle::initialFloodColor):
(WebCore::SVGRenderStyle::initialLightingColor):
(WebCore::SVGRenderStyle::fillPaintColor const):
(WebCore::SVGRenderStyle::strokePaintColor const):
(WebCore::SVGRenderStyle::stopColor const):
(WebCore::SVGRenderStyle::floodColor const):
(WebCore::SVGRenderStyle::lightingColor const):
(WebCore::SVGRenderStyle::visitedLinkFillPaintColor const):
(WebCore::SVGRenderStyle::visitedLinkStrokePaintColor const):
(WebCore::SVGRenderStyle::setFillPaint):
(WebCore::SVGRenderStyle::setStrokePaint):
(WebCore::SVGRenderStyle::setStopColor):
(WebCore::SVGRenderStyle::setFloodColor):
(WebCore::SVGRenderStyle::setLightingColor):
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/rendering/style/StyleMultiColData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/RenderSVGResource.cpp:
(WebCore::inheritColorFromParentStyleIfNeeded):
(WebCore::requestPaintingResource):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTapHighlightColor):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritCaretColor):
(WebCore::Style::colorAndSVGPaintType):
(WebCore::Style::BuilderCustom::applyValueColor):
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::colorFromPrimitiveValue const):
(WebCore::Style::BuilderState::colorFromPrimitiveValueWithResolvedCurrentColor const):
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
(WebCore::SVGFEDiffuseLightingElement::setFilterEffectAttribute):
(WebCore::SVGFEDiffuseLightingElement::createFilterEffect const):
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
(WebCore::SVGFEDropShadowElement::setFilterEffectAttribute):
(WebCore::SVGFEDropShadowElement::createFilterEffect const):
* Source/WebCore/svg/SVGFEFloodElement.cpp:
(WebCore::SVGFEFloodElement::setFilterEffectAttribute):
(WebCore::SVGFEFloodElement::createFilterEffect const):
* Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
(WebCore::SVGFESpecularLightingElement::setFilterEffectAttribute):
(WebCore::SVGFESpecularLightingElement::createFilterEffect const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::sendTapHighlightForNodeIfNecessary):

Canonical link: https://commits.webkit.org/255095@main
